train_decoder.py

```diff
@@ -0,0 +1,462 @@
+    assert train_prop + test_prop + val_prop == 1
+    num_train = round(train_prop*len(available_shards))
+    num_test = round(test_prop*len(available_shards))
+    num_val = len(available_shards) - num_train - num_test
+    assert num_train + num_test + num_val == len(available_shards), f"{num_train} + {num_test} + {num_val} = {num_train + num_test + num_val} != {len(available_shards)}"
+    train_split, test_split, val_split = torch.utils.data.random_split(available_shards, [num_train, num_test, num_val], generator=torch.Generator().manual_seed(0))
+
+    # The shard number in the webdataset file names has a fixed width. We zero pad the shard numbers so they correspond to a filename.
+    train_urls = [webdataset_base_url.format(str(shard).zfill(shard_width)) for shard in train_split]
+    test_urls = [webdataset_base_url.format(str(shard).zfill(shard_width)) for shard in test_split]
+    val_urls = [webdataset_base_url.format(str(shard).zfill(shard_width)) for shard in val_split]
+    
+    create_dataloader = lambda tar_urls, shuffle, with_text=False, is_low=False: create_image_embedding_dataloader(
+        tar_url=tar_urls,
+        num_workers=num_workers,
+        batch_size=batch_size if not is_low else 32,
+        embeddings_url=embeddings_url,
+        index_width=index_width,
+        shuffle_num = None,
+        extra_keys= ["txt"] if with_text else [],
+        shuffle_shards = shuffle,
+        resample_shards = False, 
+        img_preproc=img_preproc,
+        handler=wds.handlers.warn_and_continue
+    )
+
+    train_dataloader = create_dataloader(train_urls, shuffle_train)
+    val_dataloader = create_dataloader(val_urls, shuffle_val_test, with_text=True)
+    test_dataloader = create_dataloader(test_urls, shuffle_val_test, with_text=True, is_low=True)
+    return {
+        "train": train_dataloader,
+        "val": val_dataloader,
+        "test": test_dataloader
+    }
+
+
+def create_decoder(device, decoder_config, unets_config):
+    """Creates a sample decoder"""
+    unets = []
+    for i in range(0, len(unets_config)):
+        unets.append(Unet(
+            **unets_config[i]
+        ))
+        unets[i].to(device=device)
+    
+    decoder = Decoder(
+        unet=tuple(unets),  # Must be tuple because of cast_tuple
+        **decoder_config
+    )
+    decoder.to(device=device)
+
+    return decoder
+
+def generate_samples(decoder, dataloader, epoch, device, step, n=5, text_prepend=""):
+    """
+    Generates n samples from the decoder and uploads them to wandb
+    Consistently uses the first n image embeddings from the dataloader
+    If the dataloader returns an extra text key, that is used as the caption
+    """
+    test_iter = iter(dataloader)
+    images = []
+    with torch.no_grad():
+        decoder.eval()
+        for data in test_iter:
+            if len(data) == 3:
+                img, emb, txt = data
+            else:
+                img, emb = data
+                txt = [""] * emb.shape[0]
```

Comment:
what is this doing exactly? is this to accommodate logic between text conditioned decoder vs without?

Reply:
The dataloader allows other files to be loaded from the webdataset beyond the image and embedding. In this case, the test dataset loads caption files so that it can be logged with the sample. Since we also sample the training set at some points, the text is not always available and a default is used. The other option would be to always load the text file as well. It would not have much of an effect beyond making this part a one liner.

Reply:
ohh i see, is there anyway to detect this early on and just pass in a `has_text` boolean so it is more clear? (instead of relying on the length of the tuple returned)

Reply:
I improved this by having the dataset expose which keys it is using.
https://github.com/Veldrovive/DALLE2-pytorch/blob/61ceb19045ce3d6e77cbaabd6968e720da2dbc8f/train_decoder.py#L106-L114